### ISSUE TITLE
Question about query augmentation

### ISSUE THREAD

**Issue by jiqiujia** [stanford-futuredata/ColBERT#16]
In the paper, I notice that query augmentation is done by padding the query with [MASK] token, and doc side is not appended. While in the code, i find query and doc use the same `_encode` function. If I don't miss something, the attention mask disables the functionality of padding whatever it is. 
```
    def _encode(self, x, max_length):
        input_ids = self.tokenizer.encode(x, add_special_tokens=True, max_length=max_length)

        padding_length = max_length - len(input_ids)
        attention_mask = [1] * len(input_ids) + [0] * padding_length
        input_ids = input_ids + [103] * padding_length

        return input_ids, attention_mask
```

**Comment by okhat** [comment]
Thanks for the question!

Indeed, query augmentation is only applied to queries. The v0.1 implementation (from which your snippet is taken) does this in a short way that could possibly seem confusing at first.

The above snippet appends `103`s (i.e., `[MASK]` tokens) to queries and documents and sets the `attention_mask` to zero. This attention mask controls attention _inside_ the BERT encoder. This attention is different from ColBERT attention (MaxSim). After the BERT encoder, we mask this padding _only_ for the document side.

In contrast, for the query side, the MaxSim over `[MASK]`s is not masked. Inside the BERT encoder, setting the attention mask set to zero forces the mask tokens to be computed in the last layer, and not to influence the other token representations. Empirically, this has little to no effect on performance given plenty of training data, so the simpler implementation is in this manner.

Let me know if you have further questions!

**Comment by jiqiujia** [comment]
Thanks for your reply!
I know that doc side doesn't use the mask padding now. However, I am still confused about how mask padding makes a contribution. As I understand it, mask padding doesn't affect other tokens' representation, but will other token's representation affect mask's representation? I am not sure about this.
Besides, did you experiment how the query max_length affect the performance with mask padding?

**Comment by okhat** [comment]
This relies of an implementation detail of HuggingFace transformers: the attention mask doesn't change how the values are computed. It only affects what values are attended to.

So, to answer your question: in queries, the [MASK] positions are represented by paying attention (in the 12th layer) to all _other_ query tokens. Notice this happens only in the 12th layer by default. If you do it in all 12 layers, there's little to no difference either. You can experiment with it to see how it functions.

Also, you may want to use the v0.2 code, which should be easier to understand and otherwise more extensive.

**Comment by jiqiujia** [comment]
I still have some doubts but I think query augmentation with mask padding is a great idea, and I will do some further experiments to verify it. Thank you for your reply!

**Comment by ashantanu** [comment]
I am having trouble understanding **Query Augmentation** and why does it add so much to performance. I saw the ablation study and the difference is quite significant.
From your earlier comments, it seems like augmentation happens only in the **last layer** of ColBERT, which again confuses me about the importance of this step and  why would it contribute so much to the performance.
